torch/_logging/_internal.py

```diff
@@ -0,0 +1,426 @@
+    except ImportError:
+        return False
+
+
+def _update_log_state_from_env():
+    global log_state
+    log_setting = os.environ.get(LOG_ENV_VAR, None)
+    if log_setting is not None:
+        log_state = _parse_log_settings(log_setting)
+
+
+def _has_registered_parent(log_qname):
+    cur_log = logging.getLogger(log_qname)
+
+    registered_log_qnames = log_registry.get_log_qnames()
+
+    while cur_log.parent:
+        if cur_log.name in registered_log_qnames:
+            return True
+        cur_log = cur_log.parent
+
+    return False
+
+
+def _setup_handlers(create_handler_fn, log):
+    debug_handler = _track_handler(create_handler_fn())
+    debug_handler.setFormatter(DEFAULT_FORMATTER)
+    debug_handler.setLevel(logging.DEBUG)
+    log.addHandler(debug_handler)
+
+
+handlers = WeakSet()  # type: ignore[var-annotated]
+
+
+# mark handlers that we've created
+# so we don't modify user handlers
+def _track_handler(handler):
+    handlers.add(handler)
+    return handler
+
+
+def _is_torch_handler(handler):
+    return handler in handlers
+
+
+# clears all torch handlers on specified loggers
+def _clear_handlers(log):
+    to_remove = [handler for handler in log.handlers if _is_torch_handler(handler)]
+    for handler in to_remove:
+        log.removeHandler(handler)
+
+
+def _reset_logs():
+    # reset all registered logs
+    for log_qname in log_registry.get_log_qnames():
+        log = logging.getLogger(log_qname)
+        log.setLevel(logging.WARNING)
+        log.propagate = False
+        _clear_handlers(log)
+
+    # reset all artifact and child logs
+    for artifact_log_qname in itertools.chain(
+        log_registry.get_artifact_log_qnames(), log_registry.get_child_log_qnames()
+    ):
+        log = logging.getLogger(artifact_log_qname)
+        log.setLevel(logging.NOTSET)
+        log.propagate = True
+
+
+def _get_log_state():
+    return log_state
+
+
+def _set_log_state(state):
+    global log_state
+    log_state = state
+
+
+def _init_logs(log_file_name=None):
+    _reset_logs()
+    _update_log_state_from_env()
+
+    for log_qname, level in log_state.get_log_level_pairs():
+        log = logging.getLogger(log_qname)
+        log.setLevel(level)
+
+    # setup handlers for all registered loggers
+    for log_qname in log_registry.get_log_qnames():
+        log = logging.getLogger(log_qname)
+        _setup_handlers(
+            logging.StreamHandler,
+            log,
+        )
+
+        if log_file_name is not None:
+            _setup_handlers(
+                lambda: logging.FileHandler(log_file_name),
+                log,
+            )
+
+    # configure artifact loggers, note: this must happen last
+    # since the levels of ancestor loggers are taken into account
+    for artifact_log_qname in log_registry.get_artifact_log_qnames():
+        log = logging.getLogger(artifact_log_qname)
+        configure_artifact_log(log)
+
+
+@functools.lru_cache(None)
+def warning_once(self, *args, **kwargs):
```

Comment:
so how was this resolved at the end? The API declares as a method but it can't be used as a method, no?

Reply:
1. it has to be called `as warning_once(logger, ...)`
2. it has to be imported.

Reply:
I made a proposal below of how one way to rectify this

Reply:
I ended up going with your approach and leaving this is as. I made the call that the ux was better to patch the logger class.

Reply:
have you read my last comment before it was marked as resolved? I don't think it's OK actually - I explained why. I don't know how to find that discussion

Reply:
I mean I left this as it originally was and how you guys implemented it at HF. I patch the logger class by adding this method.

Reply:
yes, and I wrote there why I changed my mind and it's not the right thing to do for both libraries.

I need to figure out how to find that comment that disappeared.

Reply:
found it via email here it is:

https://github.com/pytorch/pytorch/pull/94858#discussion_r1139343241

Reply:
let me just re-paste it here:

--------------

actually, I was thinking about this some more. So what happens when you use 2 or more of such libraries in your application and you do:

```
import transformers # which imports its logging utils
import pytorch      # which imports its logging utils
```
and now both end up overriding `logging.Logger.warning_once` - if they are the same function we are in luck, if they aren't - then there is a race condition with the last one to be imported taking over and thus there is an undefined behavior.

So probably the clean solution that each of the components needs to do is to extend the object inside `get_logger` - `transformers` is already using a custom `get_logger` so it'd be a good place to do the overrides in. Perhaps `torch` could do the same?

Regardless of how we end up doing it, let's find a solution that we can replicate in all players so that they don't trample over each other.

Reply:
thoughts?

Reply:
I'll commit your code below, I think it's fine for what's needed.

Reply:
the pytorch repo is so big, I worry that I won't get traction on opting everyone into a custom version of get_logger for this one method. I followed the pattern with everything else: have a clear way of getting functionality (ie getArtifactLogger if artifacts are desired, and now a specific warning once functionality when that's desired) and then the basic case of vanilla logging is unmodified.

Reply:
that works - thank you for handling so many nuances, @mlazos! awesome work!

Reply:
Thanks!! I really appreciate the time and effort you put into reviewing the document and PR